### PR TITLE
fix: session refresh loop if expired token is passed in headers

### LINKED ISSUE #72: Possible refresh loop when the Authorization header is set manually to the current token if it already expired
The issue seems to be:
- the app sends a request and adds `Authorization` header manually, set to the current access token
- we start interception and create a copy of the request
- we remove the authorization header and re-add if necessary since it matches the current access token
- we get a 401
- we call refresh, which updates the tokens
- start the loop again
- we create a new copy of the original request
- we no overwrite the existing authorization header since it doesn't match what we have locally (since it updated)

The issue is similar to what was resolved here: https://github.com/supertokens/supertokens-react-native/pull/123/files#diff-08906753e0eb5ab2b91ee375361349ad892b18d17afc2e22fcdd206cd1ecb148

TODOs:
1. Please add a test that checks if this issue is present
2. Fix the issue by always removing the header if we removed it to the first iteration

### PR DESCRIPTION
**Pull request by anku255**
## Summary of change
Fixes #72 

## Related issues
- https://github.com/supertokens/supertokens-android/issues/72

## Test Plan
(Write your test plan here. If you changed any code, please provide us with clear instructions on how you verified your changes work. Bonus points for screenshots and videos!)

## Documentation changes
(If relevant, please create a PR in our [docs repo](https://github.com/supertokens/docs), or create a checklist here highlighting the necessary changes)

## Checklist for important updates
- [ ] Changelog has been updated
- [ ] `frontendDriverInterfaceSupported.json` file has been updated (if needed)
   - Along with the associated array in `app/src/main/java/com/supertokens/session/Version.java`
- [ ] Changes to the version if needed
   - In `app/build.gradle`
- [ ] Issue this PR against the latest non released version branch.
   - To know which one it is, run find the latest released tag (`git tag`) in the format `vX.Y.Z`, and then find the latest branch (`git branch --all`) whose `X.Y` is greater than the latest released tag.
   - If no such branch exists, then create one from the latest released branch.

### CODE REVIEW

File: app/src/main/java/com/supertokens/session/SuperTokensCustomHttpURLConnection.java [hunk 1]
```diff
@@ -249,9 +261,18 @@ public void setDefaultUseCaches(boolean defaultusecaches) {
...
+        // it remains removed in subsequent retries even after the session is refreshed.
+        // This prevents the use of an expired access token which would no longer match the updated local access token.
+        if (wasAuthHeaderRemovedInitially) {
+            return false;
+        }
+
+
         String accessToken = Utils.getTokenForHeaderAuth(Utils.TokenType.ACCESS, applicationContext);
         String refreshToken = Utils.getTokenForHeaderAuth(Utils.TokenType.REFRESH, applicationContext);
         if (accessToken != null && refreshToken != null && value.equals("Bearer " + accessToken)) {
```
Comment (rishabhpoddar) [inline app/src/main/java/com/supertokens/session/SuperTokensCustomHttpURLConnection.java:274]: we should check with ignore case?

File: app/src/main/java/com/supertokens/session/SuperTokensInterceptor.java [hunk 2]
```diff
@@ -46,18 +49,20 @@ private Request removeAuthHeaderIfMatchesLocalToken(Request request, Request.Bui
             String refreshToken = Utils.getTokenForHeaderAuth(Utils.TokenType.REFRESH, context);
 
             if (accessToken != null && refreshToken != null && originalHeader.equals("Bearer " + accessToken)) {
```
Comment (rishabhpoddar) [inline app/src/main/java/com/supertokens/session/SuperTokensInterceptor.java:51]: shuold use ignore case?

File: testHelpers/testapp/app/src/test/java/com/example/example/SuperTokensHttpURLConnectionTest.java [hunk 3]
```diff
@@ -1104,4 +1101,63 @@ public void doAction(HttpURLConnection con) throws IOException {
...
+        int sessionRefreshCalledCount = com.example.TestUtils.getRefreshTokenCounter();
+        if (sessionRefreshCalledCount != 0) {
+            throw new Exception("Expected session refresh endpoint to be called 0 times but it was called " + sessionRefreshCalledCount + " times");
+        }
+
+        HttpURLConnection userInfoRequestConnection = SuperTokensHttpURLConnection.newRequest(new URL(userInfoAPIURL), new SuperTokensHttpURLConnection.PreConnectCallback() {
+            @Override
+            public void doAction(HttpURLConnection con) throws IOException {
+                con.setRequestMethod("GET");
+                con.setRequestProperty("Authorization", "Bearer " +  expiredAccessToken);
```
Comment (rishabhpoddar) [inline testHelpers/testapp/app/src/test/java/com/example/example/SuperTokensHttpURLConnectionTest.java:1148]: also test the case where you are fetching the access token from storage in this function and adding it to the header, that it should work as expected.

Think of more test cases, the tests seem lacking given the significance of the change. For example, if the user adds a custom authorization bearer token, that thats the one that should be used and ours should not end up overwriting it.

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/supertokens/session/SuperTokensCustomHttpURLConnection.java (hunk 1 above)
Comment (anku255) [reply app/src/main/java/com/supertokens/session/SuperTokensCustomHttpURLConnection.java]: done

File: app/src/main/java/com/supertokens/session/SuperTokensInterceptor.java (hunk 2 above)
Comment (anku255) [reply app/src/main/java/com/supertokens/session/SuperTokensInterceptor.java]: done

File: testHelpers/testapp/app/src/test/java/com/example/example/SuperTokensHttpURLConnectionTest.java (hunk 3 above)
Comment (anku255) [reply testHelpers/testapp/app/src/test/java/com/example/example/SuperTokensHttpURLConnectionTest.java:1207]: > also test the case where you are fetching the access token from storage in this function and adding it to the header, that it should work as expected.

done

>  if the user adds a custom authorization bearer token, that thats the one that should be used and ours should not end up overwriting it.

This tests already exists. See `testThatAuthHeaderIsNotIgnoredIfItDoesntMatchTheStoredAccessToken`

**Pull request merged**